// medblocks-ui@0.0.217/dist/src/medblocks/multimedia/mediaFunction.js
function createUUID() {
    let dt = new Date().getTime();
    const uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, c => {
        const r = (dt + Math.random() * 16) % 16 | 0;
        dt = Math.floor(dt / 16);
        return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
    });
    return uuid;
}
export const supabaseStoragePlugin = {
    upload: async ({ axios, file }) => {
        const form = new FormData();
        form.append('data', file);
        const response = await axios.post(`/${createUUID()}`, form, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        const key = response.data.Key.split('/')[1];
        return key;
    },
    download: async ({ axios, key }) => {
        const response = await axios({
            url: `/${key}`,
            method: 'GET',
            responseType: 'blob',
        });
        const file = new File([response.data], 'patientData');
        return URL.createObjectURL(file);
    },
};
