// medblocks-ui@0.0.217/dist/src/medblocks/boolean/checkbox.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import '@shoelace-style/shoelace/dist/components/checkbox/checkbox';
import { customElement, html, property } from 'lit-element';
import EhrElement from '../EhrElement';
let MbCheckBox = class MbCheckBox extends EhrElement {
    constructor() {
        super(...arguments);
        this.data = undefined;
        this.disabled = false;
        this.required = false;
        this.id = 'checkbox';
    }
    _handleChange(e) {
        const checkbox = e.target;
        this.data = !!checkbox.checked;
        this._mbInput.emit();
    }
    reportValidity() {
        var _a;
        const checked = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('sl-checkbox');
        return checked.reportValidity();
    }
    render() {
        if (this.variant === 'text') {
            return html `<div>
        ${this._label()}
        <p>${this.data ? 'Yes' : 'No'}</p>
      </div>`;
        }
        return html `<sl-checkbox
      id=${this.id}
      ?required=${this.required}
      ?disabled=${this.disabled}
      ?checked=${this.data}
      ?indeterminate=${this.data == null}
      @sl-change=${this._handleChange}
      >${this.label}</sl-checkbox
    >`;
    }
};
__decorate([
    property({ type: Boolean })
], MbCheckBox.prototype, "data", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbCheckBox.prototype, "disabled", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbCheckBox.prototype, "required", void 0);
__decorate([
    property({ type: String, reflect: true })
], MbCheckBox.prototype, "id", void 0);
MbCheckBox = __decorate([
    customElement('mb-checkbox')
], MbCheckBox);
export default MbCheckBox;
