// medblocks-ui@0.0.217/dist/src/medblocks/quantity/quantity.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { css, customElement, html, property, query, state } from 'lit-element';
import '@shoelace-style/shoelace/dist/components/input/input';
import '@shoelace-style/shoelace/dist/components/menu/menu';
import '@shoelace-style/shoelace/dist/components/select/select';
import '@shoelace-style/shoelace/dist/components/icon/icon';
import '@shoelace-style/shoelace/dist/components/menu-item/menu-item';
import '@shoelace-style/shoelace/dist/components/icon-button/icon-button';
import { ifDefined } from 'lit-html/directives/if-defined';
import QuantityElement from './QuantityElement';
let MbQuantity = class MbQuantity extends QuantityElement {
    constructor() {
        super(...arguments);
        this.required = false;
        this.min = 0;
        this.hideunit = false;
        this.autocomplete = false;
        this.hoist = false;
        this.placeholder = '';
        this.id = 'quantity';
        this.hideicon = false;
        this.enablesingleunit = false;
        this.units = [];
    }
    handleChildChange() {
        var _a, _b, _c;
        this.units = [...this.querySelectorAll('mb-unit')];
        if (this.units.length === 0)
            return;
        let unitToUse = this.default;
        if ((_a = this.data) === null || _a === void 0 ? void 0 : _a.unit) {
            unitToUse = this.data.unit;
        }
        const selectedUnit = (unitToUse && this.units.find(unit => unit.unit === unitToUse)) ||
            this.units[0];
        if (selectedUnit) {
            this.min = (_b = selectedUnit.min) !== null && _b !== void 0 ? _b : null;
            this.max = (_c = selectedUnit.max) !== null && _c !== void 0 ? _c : null;
            if (this.inputElement) {
                this.inputElement.min = String(this.min || '');
                this.inputElement.max = String(this.max || '');
            }
        }
    }
    reportValidity() {
        var _a;
        const input = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('sl-input');
        return input.reportValidity();
    }
    connectedCallback() {
        super.connectedCallback();
        const observer = new MutationObserver(() => {
            this.handleChildChange();
        });
        observer.observe(this, { attributes: true, childList: true });
    }
    _disabled() {
        return (this.disabled || (this.enablesingleunit ? false : this.units.length === 1));
    }
    handleInput() {
        const magnitude = this.inputElement.value
            ? Number.parseFloat(this.inputElement.value)
            : undefined;
        const unit = this.selectElement.value || this.default;
        if (magnitude == null) {
            this.data = undefined;
        }
        else {
            this.data = {
                unit,
                magnitude,
            };
        }
        const selectedUnit = this.units.find(un => un.unit === unit);
        if (selectedUnit) {
            this.min = selectedUnit.min ? selectedUnit.min : null;
            this.max = selectedUnit.max ? selectedUnit.max : null;
            if (this.inputElement) {
                this.inputElement.min = String(this.min || '');
                this.inputElement.max = String(this.max || '');
            }
        }
        this._mbInput.emit();
    }
    get displayUnit() {
        var _a, _b;
        if (this.selectElement && this.selectElement.value) {
            return this.selectElement.value;
        }
        if ((_a = this.data) === null || _a === void 0 ? void 0 : _a.unit) {
            return (_b = this.data) === null || _b === void 0 ? void 0 : _b.unit;
        }
        if (this.default) {
            return this.default;
        }
        return '';
    }
    render() {
        var _a, _b, _c, _d, _e, _f;
        if (this.variant === 'text') {
            return html `<div>
        ${this._label()}
        <div style="display:flex;gap:0 8px">
          <p>${((_a = this.data) === null || _a === void 0 ? void 0 : _a.magnitude) || '-'}</p>
          <p>${this.hideunit ? '' : ((_b = this.data) === null || _b === void 0 ? void 0 : _b.unit) || '-'}</p>
        </div>
      </div>`;
        }
        return html `
      <sl-input
        id=${`${this.id}-magnitude`}
        class=${this.hideunit ? '' : 'margin-xs'}
        .size=${this.variant === 'small' ? 'small' : 'medium'}
        .disabled=${this.disabled}
        .step=${(_c = this.step) !== null && _c !== void 0 ? _c : 'any'}
        .required=${this.required}
        .max=${this.max}
        .min=${this.min}
        label=${ifDefined(this.label)}
        type="number"
        @sl-input=${this.handleInput}
        .value=${((_e = (_d = this.data) === null || _d === void 0 ? void 0 : _d.magnitude) === null || _e === void 0 ? void 0 : _e.toString()) || ''}
        placeholder=${this.placeholder}
        autocomplete=${this.autocomplete ? 'on' : 'off'}
      ></sl-input>
      <sl-select
        id=${`${this.id}-unit`}
        exportparts="menu"
        .disabled=${this._disabled()}
        class="${this._disabled() || this.hideicon ? 'no-icon' : ''}"
        style="${this.hideunit ? 'display: none' : ''}"
        placeholder="Select units"
        .hoist=${this.hoist}
        .value=${this.displayUnit}
        .size=${this.variant === 'small' ? 'small' : 'medium'}
        @sl-change=${this.handleInput}
      >
        ${this.units.map(unit => html `<sl-menu-item
              value=${unit.unit}
              id=${`${this.id}-unit-${unit.unit}`}
              max=${unit.max}
              min=${unit.min}
              >${unit.label}</sl-menu-item
            >`)}
      </sl-select>
      <slot style="display: none" @slotchange=${this.handleChildChange}></slot>
      <p class="print-only">${((_f = this.data) === null || _f === void 0 ? void 0 : _f.magnitude) || '-'}</p>
    `;
    }
};
MbQuantity.styles = css `
    .no-icon::part(icon) {
      display: none;
    }

    :host {
      display: flex;
      flex: 1;
      align-items: flex-end;
    }

    sl-input {
      width: 0;
      flex: 3 1 auto;
      min-width: 75px;
    }

    .margin-xs {
      margin-right: var(--sl-spacing-x-small);
    }

    sl-select {
      flex: 1 1 auto;
      width: fit-content;
      min-width: 100px;
    }
    .print-only {
      display: none;
    }

    @media print {
      .print-only {
        display: inline-block;
        margin: 0px;
        padding: 2px;
      }

      sl-input {
        display: none;
      }
      sl-select {
        display: none;
      }
    }
  `;
__decorate([
    property({ type: String, reflect: true })
], MbQuantity.prototype, "default", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbQuantity.prototype, "required", void 0);
__decorate([
    property({ type: Number, reflect: true })
], MbQuantity.prototype, "max", void 0);
__decorate([
    property({ type: Number, reflect: true })
], MbQuantity.prototype, "min", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbQuantity.prototype, "hideunit", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbQuantity.prototype, "autocomplete", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbQuantity.prototype, "hoist", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbQuantity.prototype, "disabled", void 0);
__decorate([
    property({ type: Number, reflect: true })
], MbQuantity.prototype, "step", void 0);
__decorate([
    property({ type: String, reflect: true })
], MbQuantity.prototype, "placeholder", void 0);
__decorate([
    property({ type: String, reflect: true })
], MbQuantity.prototype, "id", void 0);
__decorate([
    property({ type: Boolean })
], MbQuantity.prototype, "hideicon", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbQuantity.prototype, "enablesingleunit", void 0);
__decorate([
    state()
], MbQuantity.prototype, "units", void 0);
__decorate([
    query('sl-input')
], MbQuantity.prototype, "inputElement", void 0);
__decorate([
    query('sl-select')
], MbQuantity.prototype, "selectElement", void 0);
MbQuantity = __decorate([
    customElement('mb-quantity')
], MbQuantity);
export default MbQuantity;
