// medblocks-ui@0.0.217/dist/src/medblocks/proportion/MbProportion.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { css, html, property } from 'lit-element';
import { ifDefined } from 'lit-html/directives/if-defined';
import EhrElement from '../EhrElement';
export default class MbProportion extends EhrElement {
    constructor() {
        super(...arguments);
        this.data = undefined;
        this.required = false;
        this.autocomplete = false;
        this.hoist = false;
        this.placeholder = '';
        this.id = 'proportion';
        this.hideunit = false;
    }
    _handleChange(e) {
        const inputElement = e.target;
        if (inputElement.value === '') {
            this.data = undefined;
        }
        else {
            this.data = {
                numerator: Number.parseFloat(inputElement.value),
                denominator: this.type === 'unitary' ? 1 : 100,
                type: this.type === 'unitary' ? 1 : 2,
            };
        }
        this._mbInput.emit();
    }
    reportValidity() {
        var _a;
        const input = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('sl-input');
        return input.reportValidity();
    }
    getStep() {
        if (this.step) {
            return this.step;
        }
        if (this.type === 'unitary') {
            return '0.01';
        }
        return '';
    }
    getText() {
        var _a, _b, _c;
        if ((_a = this.data) === null || _a === void 0 ? void 0 : _a.numerator) {
            if (this.type === 'percent') {
                return `${(_b = this.data) === null || _b === void 0 ? void 0 : _b.numerator} %`;
            }
            return (_c = this.data) === null || _c === void 0 ? void 0 : _c.numerator;
        }
        return '-';
    }
    render() {
        var _a, _b;
        if (this.variant === 'text') {
            return html `<div>
        ${this._label()}
        <p>${this.getText()}</p>
      </div>`;
        }
        return html `<sl-input
        id=${`${this.id}-magnitude`}
        .required=${this.required}
        .min=${this.min}
        .max=${this.max}
        .size=${this.variant === 'small' ? 'small' : 'medium'}
        .disabled=${this.disabled}
        type="number"
        .step=${this.getStep()}
        label=${ifDefined(this.label)}
        @sl-input=${this._handleChange}
        class=${this.hideunit ? '' : 'margin-xs'}
        .value=${((_b = (_a = this.data) === null || _a === void 0 ? void 0 : _a.numerator) === null || _b === void 0 ? void 0 : _b.toString()) || ''}
        placeholder=${this.placeholder}
        autocomplete=${this.autocomplete ? 'on' : 'off'}
      >
      </sl-input>
      <sl-select
        id=${`${this.id}-unit`}
        style="${this.hideunit ? 'display: none' : ''}"
        placeholder="Select units"
        class="no-icon"
        .hoist=${this.hoist}
        value="unit"
        .size=${this.variant === 'small' ? 'small' : 'medium'}
        disabled
      >
        <sl-menu-item value="unit"
          >${this.type === 'unitary' ? '/ 1' : '%'}</sl-menu-item
        >
      </sl-select> `;
    }
}
MbProportion.styles = css `
    .no-icon::part(icon) {
      display: none;
    }

    sl-input {
      width: 0;
      flex: 3 1 auto;
      min-width: 75px;
    }

    .margin-xs {
      margin-right: var(--sl-spacing-x-small);
    }

    sl-select {
      flex: 1 1 auto;
      width: 0;
      min-width: 100px;
    }

    :host {
      display: flex;
      flex: 1;
      align-items: flex-end;
    }
  `;
__decorate([
    property({ type: Object })
], MbProportion.prototype, "data", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbProportion.prototype, "required", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbProportion.prototype, "autocomplete", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbProportion.prototype, "hoist", void 0);
__decorate([
    property({ type: String, reflect: true })
], MbProportion.prototype, "step", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MbProportion.prototype, "disabled", void 0);
__decorate([
    property({ type: String, reflect: true })
], MbProportion.prototype, "placeholder", void 0);
__decorate([
    property({ type: String, reflect: true })
], MbProportion.prototype, "id", void 0);
__decorate([
    property({ type: Boolean })
], MbProportion.prototype, "hideunit", void 0);
